go/vt/vtgate/evalengine/compiler_asm.go

```diff
@@ -2725,6 +2726,72 @@ func (asm *assembler) Fn_TRIM2(col collations.TypedCollation) {
 	}, "FN TRIM VARCHAR(SP-2) VARCHAR(SP-1)")
 }
 
+func (asm *assembler) Fn_SUBSTRING2(tt sqltypes.Type, cs types.Charset, col collations.TypedCollation) {
+	asm.adjustStack(-1)
+	asm.emit(func(env *ExpressionEnv) int {
+		str := env.vm.stack[env.vm.sp-2].(*evalBytes)
+		pos := env.vm.stack[env.vm.sp-1].(*evalInt64)
+
+		end := int64(charset.Length(cs, str.bytes))
+		if pos.i < 0 {
+			pos.i += end + 1
+		}
+		if pos.i < 1 || pos.i > end {
+			str.tt = int16(tt)
+			str.bytes = nil
+			str.col = col
+			env.vm.sp--
+			return 1
+		}
+
+		res := charset.Slice(cs, str.bytes, int(pos.i-1), int(end))
+		str.tt = int16(tt)
+		str.bytes = res
+		str.col = col
+		env.vm.sp--
+		return 1
+	}, "FN SUBSTRING VARCHAR(SP-2) INT64(SP-1)")
+}
+
+func (asm *assembler) Fn_SUBSTRING3(tt sqltypes.Type, cs types.Charset, col collations.TypedCollation) {
+	asm.adjustStack(-2)
+	asm.emit(func(env *ExpressionEnv) int {
+		str := env.vm.stack[env.vm.sp-3].(*evalBytes)
+		pos := env.vm.stack[env.vm.sp-2].(*evalInt64)
+		ll := env.vm.stack[env.vm.sp-1].(*evalInt64)
+
+		end := int64(charset.Length(cs, str.bytes))
+		if pos.i < 0 {
+			pos.i += end + 1
+		}
+		if pos.i < 1 || pos.i > end {
+			str.tt = int16(tt)
+			str.bytes = nil
+			str.col = col
+			env.vm.sp -= 2
+			return 1
+		}
+
+		if ll.i < 1 {
+			str.tt = int16(tt)
+			str.bytes = nil
+			str.col = col
+			env.vm.sp -= 2
+			return 1
```

Comment:
This is the same code in the above `if` condition too. They can be merged into one

Reply:
Fixed!